pinot-spi/src/main/java/org/apache/pinot/spi/utils/SchemaFieldExtractorUtils.java

```diff
@@ -0,0 +1,105 @@
+  public static final String MAP_KEY_COLUMN_SUFFIX = "__KEYS";
+  public static final String MAP_VALUE_COLUMN_SUFFIX = "__VALUES";
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SchemaFieldExtractorUtils.class);
+
+  /**
+   * Extracts the source fields from the schema
+   * For field specs with a transform expression defined, use the arguments provided to the function
+   * Otherwise, use the column name as is
+   * TODO: for now, we assume that arguments to transform function are in the source i.e. there's no columns which are derived from transformed columns
+   */
+  public static List<String> extract(Schema schema) {
+    Set<String> sourceFieldNames = new HashSet<>();
+    for (FieldSpec fieldSpec : schema.getAllFieldSpecs()) {
+      if (!fieldSpec.isVirtualColumn()) {
+        ExpressionEvaluator expressionEvaluator = ExpressionEvaluatorFactory.getExpressionEvaluator(fieldSpec);
+
+        if (expressionEvaluator != null) {
+          sourceFieldNames.addAll(expressionEvaluator.getArguments());
+        } else {
+          sourceFieldNames.add(fieldSpec.getName());
+        }
+      }
+    }
+    return new ArrayList<>(sourceFieldNames);
+  }
+
+  /**
+   * Validates that for a field spec with transform function, the source column name and destination column name are exclusive
+   * i.e. do not allow using source column name for destination column
+   * 1. Transform function of a field spec should not use the destination column
+   * 2. TimeFieldSpec - cannot have same name for incoming and outgoing field spec, if the specs are different
+   */
+  public static boolean validate(Schema schema) {
+    for (FieldSpec fieldSpec : schema.getAllFieldSpecs()) {
+      if (!fieldSpec.isVirtualColumn()) {
+        String column = fieldSpec.getName();
+        String transformFunction = fieldSpec.getTransformFunction();
+        if (transformFunction != null) {
+          ExpressionEvaluator expressionEvaluator = ExpressionEvaluatorFactory.getExpressionEvaluator(fieldSpec);
+          if (expressionEvaluator != null) {
+            List<String> arguments = expressionEvaluator.getArguments();
+            // output column used as input
+            if (arguments.contains(column)) {
+              LOGGER.error("The arguments of transform function: {}, should not contain the destination column: {}",
+                  transformFunction, column);
+              return false;
+            }
+          }
+        } else if (fieldSpec.getFieldType().equals(FieldSpec.FieldType.TIME)) {
+          TimeFieldSpec timeFieldSpec = (TimeFieldSpec) fieldSpec;
+          TimeGranularitySpec incomingGranularitySpec = timeFieldSpec.getIncomingGranularitySpec();
+          TimeGranularitySpec outgoingGranularitySpec = timeFieldSpec.getOutgoingGranularitySpec();
+          // different incoming and outgoing spec, but same name
+          if (outgoingGranularitySpec != null && !incomingGranularitySpec.equals(outgoingGranularitySpec)
```

Comment:
(nit) outgoingGranularitySpec will never be null because it will return incoming spec if outgoing is not set